File: solutions/go/forth/3/forth.go
package forth

import (
	"errors"
	"strconv"
	"strings"
)

func Forth(input []string) (stack []int, err error) {
	inter := newForthInterpreter()
	for _, line := range input {
		for _, token := range strings.Fields(strings.ToUpper(line)) {
			if inter.state, err = inter.state.dispatch(token); err != nil {
				return
			}
		}
	}
	stack = inter.stack
	return
}

type forthInterpreter struct {
	stack       []int
	state       forthState
	definitions map[string][]string
}

func newForthInterpreter() *forthInterpreter {
	inter := &forthInterpreter{
		stack:       make([]int, 0),
		definitions: make(map[string][]string),
	}
	inter.state = &interpretState{
		forthInterpreter: inter,
	}
	return inter
}

func (inter *forthInterpreter) interpret(token string) error {
	var err error
	if intToken, err := strconv.Atoi(token); err == nil {
		inter.stack = append(inter.stack, intToken)
		return nil
	}
	if instructions, ok := inter.definitions[token]; ok {
		for _, instruction := range instructions {
			if err = inter.interpret(instruction); err != nil {
				return err
			}
		}
		return nil
	}
	if builtin, ok := builtins[token]; ok {
		if inter.stack, err = builtin(inter.stack); err != nil {
			return err
		}
		return nil
	} else {
		return errors.New("undefined operation")
	}
}

type forthState interface {
	dispatch(token string) (forthState, error)
}
type newWordState struct {
	*forthInterpreter
}

func (state *newWordState) dispatch(token string) (forthState, error) {
	if _, err := strconv.Atoi(token); err == nil {
		return state, errors.New("illegal operation")
	}
	return &defineWordState{
		forthInterpreter: state.forthInterpreter,
		word:             token,
	}, nil
}

type defineWordState struct {
	*forthInterpreter
	word         string
	instructions []string
}

func (state *defineWordState) isDefined(token string) bool {
	if _, ok := state.definitions[token]; ok {
		return true
	}
	if _, ok := builtins[token]; ok {
		return true
	}
	_, err := strconv.Atoi(token)
	return err == nil
}

func (state *defineWordState) dispatch(token string) (forthState, error) {
	if token == ";" {
		state.definitions[state.word] = state.instructions
		return &interpretState{
			forthInterpreter: state.forthInterpreter,
		}, nil
	}
	if !state.isDefined(token) {
		return state, errors.New("undefined operation")
	}
	if predefined, ok := state.definitions[token]; ok {
		state.instructions = append(state.instructions, predefined...)
	} else {
		state.instructions = append(state.instructions, token)
	}
	return state, nil
}

type interpretState struct {
	*forthInterpreter
}

func (state *interpretState) dispatch(token string) (forthState, error) {
	if token == ":" {
		return &newWordState{
			forthInterpreter: state.forthInterpreter,
		}, nil
	}
	err := state.interpret(token)
	return state, err
}

var builtins = map[string]func([]int) ([]int, error){
	"+":    plus,
	"-":    minus,
	"*":    multiply,
	"/":    divide,
	"SWAP": swap,
	"DROP": drop,
	"DUP":  dup,
	"OVER": over,
}

func plus(stack []int) ([]int, error) {
	return binaryOperation(stack, func(a, b int) []int {
		return []int{a + b}
	})
}

func minus(stack []int) ([]int, error) {
	return binaryOperation(stack, func(a, b int) []int {
		return []int{a - b}
	})
}

func multiply(stack []int) ([]int, error) {
	return binaryOperation(stack, func(a, b int) []int {
		return []int{a * b}
	})
}

func divide(stack []int) ([]int, error) {
	if len(stack) > 0 && stack[len(stack)-1] == 0 {
		return stack, errors.New("divide by zero")
	}
	return binaryOperation(stack, func(a, b int) []int {
		return []int{a / b}
	})
}

func swap(stack []int) ([]int, error) {
	return binaryOperation(stack, func(a, b int) []int {
		return []int{b, a}
	})
}

func drop(stack []int) ([]int, error) {
	return unaryOperation(stack, func(a int) []int {
		return []int{}
	})
}

func dup(stack []int) ([]int, error) {
	return unaryOperation(stack, func(a int) []int {
		return []int{a, a}
	})
}

func over(stack []int) ([]int, error) {
	return binaryOperation(stack, func(a, b int) []int {
		return []int{a, b, a}
	})
}

func unaryOperation(stack []int, op func(int) []int) ([]int, error) {
	if len(stack) < 1 {
		return stack, errors.New("empty stack")
	}
	result := op(stack[len(stack)-1])
	stack = stack[:len(stack)-1]
	return append(stack, result...), nil
}

func binaryOperation(stack []int, op func(int, int) []int) ([]int, error) {
	if len(stack) < 1 {
		return stack, errors.New("empty stack")
	}
	if len(stack) < 2 {
		return stack, errors.New("only one value on the stack")
	}
	result := op(stack[len(stack)-2], stack[len(stack)-1])
	stack = stack[:len(stack)-2]
	return append(stack, result...), nil
}

File: solutions/kotlin/forth/2/src/main/kotlin/Forth.kt
class Forth {
    fun evaluate(vararg line: String): List<Int> = run {
        line
            .flatMap { it.split(' ') }
            .filter { it.isNotEmpty() }
            .map { it.uppercase() }
            .forEach { state = state.dispatch(it) }
        stack.toList()
    }

    private val stack = ArrayDeque<Int>()

    private var state: State = State.Interpret(this)

    private sealed class State(val outer: Forth) {
        open fun dispatch(token: String): State = when (token) {
            ":" -> NewWord(outer)
            else -> this
        }

        class NewWord(outer: Forth) : State(outer) {
            override fun dispatch(token: String): State = run {
                require(token.toIntOrNull() == null) { "illegal operation" }
                DefineWord(outer, token)
            }
        }

        class DefineWord(outer: Forth, val word: String) : State(outer) {
            val instructions = mutableListOf<String>()

            private fun isDefined(token: String): Boolean =
                token in outer.definitions || builtinFromString(token) != null || token.toIntOrNull() != null

            override fun dispatch(token: String): State = when (token) {
                ";" -> {
                    outer.definitions[word] = instructions
                    Interpret(outer)
                }

                else -> {
                    require(isDefined(token)) { "undefined operation" }
                    when (val predefined = outer.definitions[token]) {
                        null -> instructions.add(token)
                        else -> instructions.addAll(predefined)
                    }
                    this
                }
            }
        }

        class Interpret(outer: Forth) : State(outer) {
            override fun dispatch(token: String): State = when (val defaultDispatch = super.dispatch(token)) {
                this -> {
                    outer.interpret(token)
                    this
                }

                else -> defaultDispatch
            }
        }
    }

    private val definitions = mutableMapOf<String, List<String>>()

    private fun interpret(token: String): Unit = when (val intToken = token.toIntOrNull()) {
        null -> {
            when (val instructions = definitions[token]) {
                null -> {
                    val builtin = builtinFromString(token)
                    require(builtin != null) { "undefined operation" }
                    interpretBuiltin(builtin)
                }

                else -> instructions.forEach { interpret(it) }
            }
        }

        else -> stack.addLast(intToken)
    }

    enum class Builtin(val token: String) {
        PLUS("+"), MINUS("-"), MULTIPLY("*"), DIVIDE("/"),
        DUP("DUP"), OVER("OVER"), DROP("DROP"), SWAP("SWAP")
    }

    companion object {
        val BUILTINS = Builtin.values()
        private fun builtinFromString(token: String): Builtin? = BUILTINS.find { it.token == token }
    }

    private fun interpretBuiltin(builtin: Builtin): Unit = when (builtin) {
        Builtin.PLUS -> {
            binaryValidation()
            val last = stack.removeLast()
            stack[stack.lastIndex] += last
        }

        Builtin.MINUS -> {
            binaryValidation()
            val last = stack.removeLast()
            stack[stack.lastIndex] -= last
        }

        Builtin.MULTIPLY -> {
            binaryValidation()
            val last = stack.removeLast()
            stack[stack.lastIndex] *= last
        }

        Builtin.DIVIDE -> {
            binaryValidation()
            require(stack.last() != 0) { "divide by zero" }
            val last = stack.removeLast()
            stack[stack.lastIndex] /= last
        }

        Builtin.SWAP -> {
            binaryValidation()
            stack[stack.lastIndex - 1] = stack[stack.lastIndex]
                .also { stack[stack.lastIndex] = stack[stack.lastIndex - 1] }
        }

        Builtin.DROP -> {
            unaryValidation()
            stack.removeLast()
            Unit
        }

        Builtin.DUP -> {
            unaryValidation()
            stack.addLast(stack.last())
        }

        Builtin.OVER -> {
            binaryValidation()
            stack.addLast(stack[stack.lastIndex - 1])
        }
    }

    private fun unaryValidation() = run {
        require(stack.isNotEmpty()) { "empty stack" }
    }

    private fun binaryValidation() = run {
        unaryValidation()
        require(stack.size > 1) { "only one value on the stack" }
    }
}
